File: frontend/student.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Student, StudentDTO } from '../models/student.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class StudentService {
  private apiUrl = `${environment.apiUrl}/students`;

  constructor(private http: HttpClient) {}

  getAllStudents(): Observable<Student[]> {
    return this.http.get<Student[]>(this.apiUrl);
  }

  getStudentById(id: number): Observable<Student> {
    return this.http.get<Student>(`${this.apiUrl}/${id}`);
  }

  getStudentByMatricule(matricule: string): Observable<Student> {
    return this.http.get<Student>(`${this.apiUrl}/matricule/${matricule}`);
  }

  getStudentsByDepartment(department: string): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.apiUrl}/department/${department}`);
  }

  getStudentsByLevel(level: string): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.apiUrl}/level/${level}`);
  }

  createStudent(student: StudentDTO): Observable<Student> {
    return this.http.post<Student>(this.apiUrl, student);
  }

  updateStudent(id: number, student: StudentDTO): Observable<Student> {
    return this.http.put<Student>(`${this.apiUrl}/${id}`, student);
  }

  deleteStudent(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  searchStudents(keyword: string): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.apiUrl}/search?keyword=${keyword}`);
  }

  getActiveStudents(): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.apiUrl}/active/list`);
  }
}

File: backend/student-service/src/main/java/com/isj/studentmanagement/service/StudentService.java
package com.isj.studentmanagement.service;

import com.isj.studentmanagement.model.Student;
import com.isj.studentmanagement.repository.StudentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class StudentService {
    private final StudentRepository studentRepository;

    public List<Student> getAllStudents() {
        return studentRepository.findAll();
    }

    public Student getStudentById(Long id) {
        return studentRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Student not found with id: " + id));
    }

    public Student getStudentByMatricule(String matricule) {
        return studentRepository.findByMatricule(matricule)
            .orElseThrow(() -> new RuntimeException("Student not found with matricule: " + matricule));
    }

    public List<Student> getStudentsByDepartment(String department) {
        return studentRepository.findByDepartment(department);
    }

    public List<Student> getStudentsByLevel(String level) {
        return studentRepository.findByLevel(level);
    }

    public Student createStudent(Student student) {
        if (studentRepository.existsByMatricule(student.getMatricule())) {
            throw new RuntimeException("Student with matricule already exists: " + student.getMatricule());
        }
        if (studentRepository.existsByEmail(student.getEmail())) {
            throw new RuntimeException("Student with email already exists: " + student.getEmail());
        }
        return studentRepository.save(student);
    }

    public Student updateStudent(Long id, Student studentDetails) {
        Student student = getStudentById(id);

        if (!student.getEmail().equals(studentDetails.getEmail()) &&
            studentRepository.existsByEmail(studentDetails.getEmail())) {
            throw new RuntimeException("Email already in use: " + studentDetails.getEmail());
        }

        student.setFirstName(studentDetails.getFirstName());
        student.setLastName(studentDetails.getLastName());
        student.setEmail(studentDetails.getEmail());
        student.setPhone(studentDetails.getPhone());
        student.setDateOfBirth(studentDetails.getDateOfBirth());
        student.setGender(studentDetails.getGender());
        student.setAddress(studentDetails.getAddress());
        student.setDepartment(studentDetails.getDepartment());
        student.setLevel(studentDetails.getLevel());
        student.setActive(studentDetails.getActive());

        return studentRepository.save(student);
    }

    public void deleteStudent(Long id) {
        Student student = getStudentById(id);
        studentRepository.delete(student);
    }

    public List<Student> searchStudents(String keyword) {
        return studentRepository.findByFirstNameContainingIgnoreCaseOrLastNameContainingIgnoreCase(keyword, keyword);
    }

    public List<Student> getActiveStudents() {
        return studentRepository.findByActive(true);
    }
}
